src/Serializer/Mapping/Loader/PropertyMetadataLoader.php

```diff
@@ -0,0 +1,159 @@
+                $classMetadata->setClassDiscriminatorMapping(new ClassDiscriminatorMapping(
+                    $attribute->getTypeProperty(),
+                    $attribute->getMapping()
+                ));
+                continue;
+            }
+
+            if ($attribute instanceof Groups) {
+                $classGroups = $attribute->getGroups();
+
+                continue;
+            }
+
+            if ($attribute instanceof Context) {
+                $classContextAnnotation = $attribute;
+            }
+        }
+
+        foreach ($refl->getProperties() as $reflProperty) {
+            foreach ($reflProperty->getAttributes(ApiProperty::class) as $propAttr) {
+                $this->addAttributeMetadata($propAttr->newInstance()->withProperty($reflProperty->name), $attributes);
+            }
+        }
+
+        foreach ($refl->getMethods() as $reflMethod) {
+            foreach ($reflMethod->getAttributes(ApiProperty::class) as $methodAttr) {
+                $this->addAttributeMetadata($methodAttr->newInstance()->withProperty($reflMethod->getName()), $attributes);
+            }
+        }
+
+        foreach ($this->propertyNameCollectionFactory->create($classMetadata->getName()) as $propertyName) {
+            if (!isset($attributesMetadata[$propertyName])) {
+                $attributesMetadata[$propertyName] = new AttributeMetadata($propertyName);
+                $classMetadata->addAttributeMetadata($attributesMetadata[$propertyName]);
+            }
+
+            foreach ($classGroups as $group) {
+                $attributesMetadata[$propertyName]->addGroup($group);
+            }
+
+            if ($classContextAnnotation) {
+                $this->setAttributeContextsForGroups($classContextAnnotation, $attributesMetadata[$propertyName]);
+            }
+
+            if (!isset($attributes[$propertyName])) {
+                continue;
+            }
+
+            foreach ($attributes[$propertyName] as $attr) {
+                if ($attr instanceof Groups) {
+                    foreach ($attr->getGroups() as $group) {
+                        $attributesMetadata[$propertyName]->addGroup($group);
+                    }
+                } elseif ($attr instanceof MaxDepth) {
+                    $attributesMetadata[$propertyName]->setMaxDepth($attr->getMaxDepth());
+                } elseif ($attr instanceof SerializedName) {
+                    $attributesMetadata[$propertyName]->setSerializedName($attr->getSerializedName());
+                } elseif ($attr instanceof SerializedPath) {
+                    $attributesMetadata[$propertyName]->setSerializedPath($attr->getSerializedPath());
+                } elseif ($attr instanceof Ignore) {
+                    $attributesMetadata[$propertyName]->setIgnore(true);
+                } elseif ($attr instanceof Context) {
+                    $this->setAttributeContextsForGroups($attr, $attributesMetadata[$propertyName]);
+                }
```

Comment:
`match`?

(I guess this comes from Symfony, if you want to keep the code as-is to facilitate backporting the code if it changes upstream it's ok too, but please add a comment linking to the original source in this case)